### PR TITLE
Machine ID: Introduce BotService

### LINKED ISSUE #10477: Comprehensive audit log events for the certificate renewal bot
### What
Audit log events should be emitted for various actions performed by / relating to certificate bots. Per the [RFD](https://github.com/gravitational/teleport/pull/7986) the following events should be emitted:
- a new bot user is created
- a new renewable certificate is issued for the first time
- a certificate is renewed
- a bot is locked
- a bot is removed (likely due to failed heartbeating or expired certificates)
- a certificate generation counter conflict is detected (certificate possibly compromised)

### LINKED ISSUE #33808: Machine ID: Introduce Bot resource and RPCs
Currently, Bots in Teleport do not exist as a distinct resource. To create a 
Bot, the CreateBot RPC is invoked and this creates a user and role. This RPC 
does not treat a Bot as an entity with a spec, as we do for most resources in
Teleport. It is also not possible to update an existing bot (although, it is
possible to modify the constituent role and user).

```proto
// CreateBotRequest is used to create a bot User and associated resources.
message CreateBotRequest {
  // Name is the name of the bot, i.e. the unprefixed User name.
  string Name = 1 [(gogoproto.jsontag) = "name"];

  // TTL is the desired TTL for the token if one is created. If unset, a
  // server default is used.
  int64 TTL = 2 [
    (gogoproto.jsontag) = "ttl",
    (gogoproto.casttype) = "Duration"
  ];

  // TokenID is an optional token name of an EC2/IAM join token should be
  // used. If unset, a new random token is created and its name returned.
  string TokenID = 3 [(gogoproto.jsontag) = "token_id"];

  // Roles is a list of roles the created bot should be allowed to assume
  // via role impersonation.
  repeated string Roles = 4 [(gogoproto.jsontag) = "roles"];

  // Traits are used to populate role variables. These will propagate to
  // role impersonated certificates generated by the bot.
  wrappers.LabelValues Traits = 5 [
    (gogoproto.nullable) = false,
    (gogoproto.jsontag) = "traits,omitempty",
    (gogoproto.customtype) = "github.com/gravitational/teleport/api/types/wrappers.Traits"
  ];
}
```

This causes the following challenges:

- As there is no distinct Bot resource, our existing tools for managing
  resources do not work. E.G You cannot `tctl create -f bot.yaml`. As Machine ID
  grows more complex, there will need to be more configuration options - these 
  will not be natural to provide via commandline
- As it is not possible to modify an existing Bot, Day 1/2 operations are more
  challenging. Operators must either perform surgery on the constituent user
  and role, or, they must delete and recreate the Bot. This also means our
  Terraform provider is much more destructive than is necessary.

As Machine ID develops, the need for a distinct resource has become more
apparent. 

## Initial Work

The goal is to provide a seamless transition for users and to avoid biting off
more than we can chew with this. Therefore, this work will be broken down into
distinct, individually beneficial stages.

The first stage introduces a Bot resource type and RPCs for the management of this resource type.

The actual spec is simple - simpler than the existing `CreateBot` RPC params - this is because we'll be omitting some behaviour (e.g the automatic token creation) and requiring that the client completes this behaviour instead.

```yaml
kind: bot
version: v1
metadata:
  name: github-k8s-deployer
spec:
  // roles are configured as impersonatable roles on the Bot Role
  roles:
  - editor
  // traits is the same as on the User resource, and is just propagated straight down to Bot User.
  traits:
    logins:
    - root
```

RPCs:
- [Create/Upsert]Bot
- ListBots
- GetBot
- DeleteBot

Rather than persist the Bot resource to the backend, this initial work will create and update the Bot User and Bot Role when the Bot resource is created/updated - essentially syncing on any action.

It is important that this conversion should be compatible with any existing Bots that have already been created.

Terraform and `tctl` should be updated to support the new resource.

### PR DESCRIPTION
**Pull request by strideynet**
Closes https://github.com/gravitational/teleport/issues/33808
Closes https://github.com/gravitational/teleport/issues/10477

Quite a few deprecations in this one. Deprecated RPCs and perms now emit a warning log message which should encourage most folks to migrate prior to v16.0.0.

Fairly large diff - in part due to about 1800 lines of tests that expand the coverage of bot related code significantly. Some of these tests are for the legacy/deprecated endpoints that weren't covered previously but I wanted to make sure they behaved properly in their deprecated state.

changelog: SEE PR

Machine ID Bots can now be managed as yaml with the typical `tctl` commands, e.g:

- `tctl get bot foo`
- `tctl get bots`
- `tctl create -f bot.yaml`

```yaml
kind: bot
version: v1
metadata:
  name: test
spec:
  roles:
  - editor
  traits:
  - name: logins
    values:
    - root
```

The ability to specifically control access to managing Bots has been added to Teleport's RBAC engine. Users managing Bots should now be explicitly granted permissions to manage the `bot` resource. eg:

```yaml
spec:
  allow:
    rules:
    - resources:
      - bot
      verbs:
      - list
      - create
      - read
      - update
      - delete
```

The previous required permissions, a combination of access to `user` and `role`, will no longer grant access to Bot RPCs in Teleport 16.0.0.

Additionally, three new audit events are now emitted:

- `bot.create`
- `bot.update`
- `bot.delete`

For those using the Teleport API, the following RPCs are deprecated and will stop functioning in Teleport 16.0.0:

- `proto.AuthService.CreateBot` - switch to `machineid.v1.BotService.CreateBot`
- `proto.AuthService.DeleteBot` - switch to `machineid.v1.BotService.DeleteBot`
- `proto.AuthService.GetBotUsers` - switch to`machineid.v1.BotService.ListBots`

### CODE REVIEW

File: tool/tctl/common/bots_command.go
Comment (strideynet) [inline tool/tctl/common/bots_command.go:149]: The behaviour of `tctl bots ls` effectively changes from listing bot users to listing bots. I don't know of any consumers who are depending on the old behaviour and whilst this is effectively the same, it is a breaking change if someone was depending on fields in the table view having certain values. I don't think the previous info was too useful since it just informed you of the Bot role rather than the roles the bot can impersonate, but thought this was worth highlighting.

File: lib/auth/init.go
Comment (strideynet) [inline lib/auth/init.go:1411]: FYI @codingllama - this may be of interest to you in terms of how the Adapter works in practice.

Comment (codingllama) [inline lib/auth/init.go:1411]: Maybe:

```go
	for _, resource := range resources {
		var res any = resource

		// Unwrap "new style" resources.
		// We always want to switch over the underlying type.
		if w, ok := res.(interface{ Unwrap() types.Resource153 }); ok {
			res = w.Unwrap()
		}

		switch r := res.(type) {
		case types.ProvisionToken:
			err = service.Provisioner.UpsertToken(ctx, r)
		case types.User:
			err = services.ValidateUserRoles(ctx, r, service.Access)
			if err != nil {
				return trace.Wrap(err)
			}
			_, err = service.Identity.UpsertUser(ctx, r)
		case types.Role:
			_, err = service.Access.UpsertRole(ctx, r)
		case types.ClusterNetworkingConfig:
			err = service.ClusterConfiguration.SetClusterNetworkingConfig(ctx, r)
		case types.AuthPreference:
			err = service.ClusterConfiguration.SetAuthPreference(ctx, r)
		case *machineidv1pb.Bot:
			_, err := machineidv1.UpsertBot(ctx, service, r)
		default:
			return trace.NotImplemented("cannot apply resource of type %T", resource)
		}
		if err != nil {
			return trace.Wrap(err)
		}
	}
```

File: lib/auth/init.go
Comment (codingllama) [inline lib/auth/init.go:1417]: Let err be handled below, like the others?

```suggestion
				_, err = machineidv1.UpsertBot(ctx, service, r)
```

File: lib/auth/init.go
Comment (codingllama) [reply lib/auth/init.go:1411]: Thanks for copying me. If this is the only Unwrap then I suppose I'm somewhat OK with it, but let me know if you need to add more.

It also looks like a switch over kind is possible. The semantics might be better, but the type assertions are still necessary.

This could also be candidate for migration to a smaller interface, as it doesn't really seem to need anything apart from Kind. FYI @rosstimothy.

**Review by codingllama — COMMENTED** [review]
Drive-by review, just regarding that one comment.

File: lib/auth/init.go
Comment (strideynet) [reply lib/auth/init.go]: effb495686

File: lib/auth/init.go
Comment (strideynet) [reply lib/auth/init.go]: effb495686

File: lib/auth/init.go
Comment (strideynet) [reply lib/auth/init.go]: Thanks Alan.

> This could also be candidate for migration to a smaller interface, as it doesn't really seem to need anything apart from Kind

I do agree - this is why one of my suggestions is that instead of having `types.ResourceMetadata`, we actually probably need something like `types.ResourceVersionKind`.

Comment (codingllama) [reply lib/auth/init.go]: I'm perfectly OK with smaller, impromptu interfaces for cases like this. I don't think we should export all permutations of resource-like interfaces, like the various ResourceWithFoo we have now. ResourceMetadata is something you can use as a proxy for "this is a resource", while making few assumptions about it.

File: api/client/client.go
Comment (ravicious) [inline api/client/client.go:1146]:
```suggestion
// Deprecated: use [machineidv1pb.BotService.CreateBot] instead.
```

File: api/proto/teleport/legacy/client/proto/authservice.proto
Comment (ravicious) [inline api/proto/teleport/legacy/client/proto/authservice.proto:2648]: How can I ensure that a doc link like this is valid?

This comment ends up in `api/client/proto/authservice.pb.go`. I was under the impression that if `machineid.v1` is not directly imported in that file, then a doc link must use the full import path, which I assume would be `github.com/gravitational/teleport/…`. Isaiah pointed it out in one of my PRs but I've never tried to actually verify if my doc links are correct.

https://tip.golang.org/doc/comment#doclinks

I know it doesn't matter much as most of the time those "doc links" will not end up in any docs and are just for internal reference. Still, it's been bugging me since it was pointed out to me so I figured maybe you know the answer.

**Review by ravicious — COMMENTED** [review]
Gave it just a cursory look as I'm not the best person to review this, but I see that you already requested a review from Tim.

File: lib/auth/machineid/machineidv1/bot_service.go
Comment (ibeckermayer) [inline lib/auth/machineid/machineidv1/bot_service.go:389]: Throughout the codebase we have these `UpdateX` functions which semantically mean "if this resource already exists, overwrite it completely (a-la `Upsert`), otherwise error". AFAICT it's very unusual to need this semantic:
If you're calling `UpdateX` it almost always means that you've just called `GetX`, changed a field or two, and are now updating the object. This makes the check for "if this resource already exists" superfluous -- you already know it exists because you just got it with `GetX`.

Note that HTTP doesn't have any such semantic:
- `CREATE`: `CreateX`
- `GET`: `GetX`
- `DELETE`: `DeleteX`
- `PUT`: `UpsertX`
- `PATCH`: no equivalent except an odd `UpdateX` like `UpdateRemoteCluster`
- `POST`: according to [the RFC](https://datatracker.ietf.org/doc/html/rfc7231#section-4.3.3):
    
    > The POST method requests that the target resource process the
   representation enclosed in the request according to the resource's
   own specific semantics.
   
   Arguably our `UpdateX` is this miscellaneous type, however as I pointed out previously it's an odd semantic to implement.
   
Given this line of reasoning, I think we should aim to avoid adding these `UpdateX` by default. Usually an `Upsert` seems to do the job. 

One loose end I have on this topic is: how do `Upsert`s behave with the new optimistic locking behavior? If a resource has a `Revision`, will it be checked and rejected if it doesn't match what's in the database on an `Upsert`? If yes, then `Upsert` already implements the current `Update` semantic (and `Update`s can be replaced with `Upsert`s). If that's not how it works, and `Upserts` just ignore `Revision`, then the current `Update`s could be seen as means of distributed synchronization to prevent a case like

1. My code calls `GetX`
2. Some other part of the system calls `DeleteX` on that object
3. My code updates a field and calls `UpdateX` on that object

However I don't think this was an intentional design pattern.

cc @rosstimothy @espadolini

File: tool/tctl/common/bots_command.go
Comment (ibeckermayer) [reply tool/tctl/common/bots_command.go:150]: Could be a bullet point in the changelog

File: tool/tctl/common/collection.go
Comment (ibeckermayer) [inline tool/tctl/common/collection.go:1154]: `verbose` is unused

File: tool/tctl/common/bots_command.go
Comment (timothyb89) [inline tool/tctl/common/bots_command.go:378]: One breaking change here, previously we would output the token and join method in a machine readable way which was useful for creating bots via script:
```
$ tctl bots add test --roles=access --format=json
{
  "user_name": "bot-test",
  "role_name": "bot-test",
  "token_id": "foo",
  "ttl": 3599994659626,
  "join_method": "token"
}
```

Now we just dump the bot object, which does not have any info about the join token or method:
```
$ tctl bots add test --roles=access --format=json                                                                                                                                                                                               
{
  "kind": "bot",
  "version": "v1",
  "metadata": {
    "name": "test"
  },
  "spec": {
    "traits": [
      {
        "name": "logins"
      }
    ]
  },
  "status": {
    "user_name": "bot-test",
    "role_name": "bot-test"
  }
}
```

I don't know whether or not anyone used this - and it mostly only applies to static tokens which are a pain - but there's not a great substitute for creating bots via script now since any auto generated tokens get eaten.

(I'm not super familiar with the operator's embedded tbot but it might be worth making sure it doesn't depend on tctl's output, since it uses token based joining)

File: api/proto/teleport/machineid/v1/bot.proto
Comment (timothyb89) [inline api/proto/teleport/machineid/v1/bot.proto:67]: Morbidly curious... what is `role_role` reserved for?

**Review by timothyb89 — COMMENTED** [review]
Just a few comments after playing with the branch for a while. The implementation turned out pretty nice! The separate service and RFD 153 resources make this much tidier than I remember.

I didn't notice any compatibility issues but haven't yet tested the Terraform provider integration (which probably could use an update anyway, now that we have sane bot objects).

File: lib/auth/machineid/machineidv1/bot_service.go
Comment (strideynet) [reply lib/auth/machineid/machineidv1/bot_service.go:675]: Probably a good idea since it looks a little odd otherwise 😓

File: lib/auth/machineid/machineidv1/bot_service.go
Comment (strideynet) [reply lib/auth/machineid/machineidv1/bot_service.go:665]: Sure - sounds good

File: api/proto/teleport/machineid/v1/bot.proto
Comment (strideynet) [reply api/proto/teleport/machineid/v1/bot.proto:67]: My previous PR which added the protos had a mistake that I only noticed when I got around to this PR - the backwards compatibility linter requires that if I remove a field, I reserve it, to prevent a field with the same name being created in future.

File: tool/tctl/common/bots_command.go
Comment (strideynet) [reply tool/tctl/common/bots_command.go:378]: Good catch, I'll modify this to preserve the old JSON

File: tool/tctl/common/collection.go
Comment (strideynet) [reply tool/tctl/common/collection.go:1154]: The signature of this method must include the `verbose` parameter in order to match the `resourceCollection` interface.

File: api/proto/teleport/legacy/client/proto/authservice.proto
Comment (strideynet) [reply api/proto/teleport/legacy/client/proto/authservice.proto:2648]: Ah - I suppose it's a little odd here since my intention was to "comment on the proto" and therefore here I've referred to the proto service rather than the generated code.

File: api/client/client.go
Comment (strideynet) [reply api/client/client.go]: 238b8115de

File: lib/auth/machineid/machineidv1/bot_service.go
Comment (strideynet) [reply lib/auth/machineid/machineidv1/bot_service.go:389]: Fwiw: my implementation here is following the recently proposed RFD 153 on the design of resource APIs. So it may be worth following up with some of these concerns there - https://github.com/gravitational/teleport/pull/34103 .

I believe the long term goal is to try and eliminate the Upsert RPCs and encourage the use of the more specific Create/Update RPCs. But, right now, several things rely on the ability to upsert (e.g `tctl create -f` / the terraform provider / the kubernetes operator). Once we reach a point where all of those no longer rely on this behaviour, then I think we can almost globally deprecate Upsert RPCs.

> One loose end I have on this topic is: how do Upserts behave with the new optimistic locking behavior?

IIRC the behaviour of Upsert RPCs has not been changed. Optimistic locking is entirely ignored.

I believe that in addition to this, Update RPCs will be encouraged to support update masks where appropriate. At least in this case, my UpdateBot RPC also implements an `update_mask` field which allows selective modification of the roles or traits associated with the Bot.

File: lib/auth/init.go
Comment (strideynet) [reply lib/auth/init.go]: 07a20d9db9

File: lib/auth/machineid/machineidv1/bot_service.go
Comment (strideynet) [reply lib/auth/machineid/machineidv1/bot_service.go:96]: As Bots currently map onto a User/Role, I wanted to be cautious in terms of issues that could arise if cache propagation takes its time and changes are made in rapid succession to a Bot. I've kept the methods available in the cache interface for when the a purely read only RPC is being called. Eventually, these would probably be removed from the Backend interface when Bots become a "real" resource stored within the backend.

File: lib/auth/machineid/machineidv1/bot_service.go
Comment (strideynet) [reply lib/auth/machineid/machineidv1/bot_service.go:694]: 0c5a9aee39

File: lib/auth/machineid/machineidv1/bot_service.go
Comment (espadolini) [reply lib/auth/machineid/machineidv1/bot_service.go:416]: Upsert is by definition unconditional, and it's required (at least as a primitive operation, it doesn't have to be exposed through grpc necessarily) for the cache to work.

The old `Update` primitive backend operation has no right to exist, and it has the semantics that @ibeckermayer mentioned. `ConditionalUpdate` is the one that you want to use, and you use exactly like that. You get the resource, you make some decisions and calculate the new resource, you `ConditionalUpdate` the resource with a constraint on the revision you got, and your update succeeds if and only if nothing has written the underlying backend item.

Whether the revision should be exposed to the client over grpc is a choice that should be done depending on the resource type and the semantics we want to give to each operation; it might make sense sometimes to expose an "update" function that lets the caller update some fields with no regard to what was there in the first place, but even in that case the auth server should 100% use the conditional update operation internally.

Comment (espadolini) [reply lib/auth/machineid/machineidv1/bot_service.go:416]: (and yeah, "i just read the item successfully so the item exists" is 100% bogus and it's a pattern that we should start removing from Teleport whenever possible)

File: tool/tctl/common/bots_command.go
Comment (strideynet) [reply tool/tctl/common/bots_command.go]: [d552a14](https://github.com/gravitational/teleport/pull/35441/commits/d552a14533cd3e5311bba8df6ce72fb7adc11553)

File: api/proto/teleport/machineid/v1/bot.proto
Comment (timothyb89) [reply api/proto/teleport/machineid/v1/bot.proto:67]: Ah, got it, thanks!

**Review by timothyb89 — APPROVED** [review]
Looks good, I'm excited to finally see this!

File: lib/auth/machineid/machineidv1/bot_service.go
Comment (ibeckermayer) [reply lib/auth/machineid/machineidv1/bot_service.go:416]: #### @strideynet 
> It was my understanding that this was the entire motivation behind adding optimistic locking to the update RPC.

I was referring to the state of things pre-optimistic-locking.

> I believe if anything long term goal is to do the opposite of what you suggest, and eliminate upsert in favour of create and update. Right now, several things rely on the ability to upsert (e.g tctl create -f / the terraform provider / the kubernetes operator). Once we reach a point where all of those no longer rely on this behaviour, then I think we can almost globally deprecate upsert RPCs.

Seems like an odd goal... Is `create -f` not the user explicitly asking for an upsert? I can imagine implementing it with a get + update, but I'm not currently seeing what the point of that would be.

#### @espadolini 

> Upsert is by definition unconditional

Granted, dumb question.

> The old Update primitive backend operation has no right to exist, and it has the semantics that @ibeckermayer mentioned. ConditionalUpdate is the one that you want to use, and you use exactly like that. You get the resource, you make some decisions and calculate the new resource, you ConditionalUpdate the resource with a constraint on the revision you got, and your update succeeds if and only if nothing has written the underlying backend item since you read it.

Is this not essentially what `Update` has become, presuming it now uses the optimistic locking mechanism? 

I can see several plausible update-like semantics:

1. update the entire object irrespective of revision (error if object dne)
2. update the entire object on a specific revision (error if revision/object dne)
3. patch a few fields irrespective of revision (error if object dne)
4. patch a few fields on a specific revision (error if revision/object dne)
5. write an entire object irrespective of revision/existence (upsert)

1. is the status quo 
2. is presumably the primary reason we implemented optimistic locking 
3. we have in a few places (HTTP patch semantic)
4. seems practically useless -- if we have the object's revision, that typically means we have the object, so we may as well just use 2
5. seems useful to me, although there seems to be some discussion of getting rid of it

Comment (ibeckermayer) [reply lib/auth/machineid/machineidv1/bot_service.go:416]: Thanks for pointing out that PR @strideynet, this whole discussion is more suited to that RFD so let's continue it over there: https://github.com/gravitational/teleport/pull/34103#discussion_r1428615401

Comment (strideynet) [reply lib/auth/machineid/machineidv1/bot_service.go:416]: @ibeckermayer do you think this blocks this PR moving forward ? I'm a little concerned about how long discussing this will run and I have some time pressure to get this merged as it's blocking other work.

Comment (ibeckermayer) [reply lib/auth/machineid/machineidv1/bot_service.go:416]: Not necessarily, I can review the rest and we can just consider this pending. 

Since you did go through the trouble of adding a field mask, it's worth following [this line of inquiry](https://github.com/gravitational/teleport/pull/34103/files/b5916b1a93a818a1b7d0667beaaf0c905ae19655#r1430639369) through to its conclusion.

File: tool/tctl/common/bots_command.go
Comment (ibeckermayer) [inline tool/tctl/common/bots_command.go:403]: Edgiest of edge cases, but at this point the bot has been created, so getting an error might confuse the customer about the state of their cluster. Consider acknowledging that the bot was created in this error message.

File: tool/tctl/common/bots_command.go
Comment (ibeckermayer) [inline tool/tctl/common/bots_command.go:291]: Perhaps deserving of a `// DELETE in X.Y`

File: tool/tctl/common/bots_command.go
Comment (ibeckermayer) [inline tool/tctl/common/bots_command.go:424]: Could use a similar note about checking for versioning as `AddBot` has.

**Review by ibeckermayer — APPROVED** [review]

File: lib/auth/machineid/machineidv1/bot_service.go
Comment (ibeckermayer) [reply lib/auth/machineid/machineidv1/bot_service.go:96]: Maybe worth a note somewhere, your call.

File: tool/tctl/common/bots_command.go
Comment (strideynet) [reply tool/tctl/common/bots_command.go:424]: Yup - I'll do this tomorrow.

File: tool/tctl/common/bots_command.go
Comment (strideynet) [reply tool/tctl/common/bots_command.go:403]: Yeah, although this is the old message as well, I'll go ahead and update it since I agree this is unclear.

File: tool/tctl/common/bots_command.go
Comment (strideynet) [reply tool/tctl/common/bots_command.go:425]: e472f48327

File: tool/tctl/common/bots_command.go
Comment (strideynet) [reply tool/tctl/common/bots_command.go:292]: e472f48327

File: tool/tctl/common/bots_command.go
Comment (strideynet) [reply tool/tctl/common/bots_command.go]: 3b27833c7b

**Comment by strideynet** [comment]
Flaky Test Detector is failing due to test length - it looks like the `TestAdminActionMFA` test takes about 7.67 seconds to run on my mac, which at 100 repetitions passes the 10 minute threshold. As far as I can tell, the changes I've made are only adding about 200-400ms to this test. CC @Joerger

**Comment by Joerger** [comment]
@strideynet This test will be skipped by the flaky test detector momentarily - https://github.com/gravitational/teleport/pull/35882

**Comment by strideynet** [comment]
Merged `master` in to pick up joerger's fix for the flaky-test failure

File: lib/auth/machineid/machineidv1/bot_service.go
Comment (zmb3) [inline lib/auth/machineid/machineidv1/bot_service.go:58]: Do we need to remove `/` as well?

**Review by zmb3 — APPROVED** [review]

File: api/types/user.go
Comment (strideynet) [reply api/types/user.go:509]: The Bot resource is still just transformed into a User/Role - eventually we'll remove this but that'll involve delving into the core of Teleport's RBAC 😓 

This change here is just to make this a little less fragile. I'm not quite sure why we used the Generation label here - it stays at "0" for bots using delegated joining and could be removed in future. The `BotLabel` is a much better indicator of a user/role being linked to a bot.

File: lib/auth/machineid/machineidv1/bot_service.go
Comment (strideynet) [reply lib/auth/machineid/machineidv1/bot_service.go:167]: I suppose 😂 - I have a bad habit of using a `switch` by default for validating requests at the start of a gRPC handler from a codebase I previously worked on.

File: lib/auth/machineid/machineidv1/bot_service.go
Comment (strideynet) [reply lib/auth/machineid/machineidv1/bot_service.go:58]: At the moment, this has just moved the previous implementation of `BotResourceName` from `lib/auth` into `lib/auth/machineid/machineidv1` - I do agree that maybe we ought to replace `/` but I feel it's out of scope for this PR and I don't want to introduce additional risk by changing this.

File: lib/auth/machineid/machineidv1/bot_service.go
Comment (strideynet) [reply lib/auth/machineid/machineidv1/bot_service.go:167]: 83272c932d

**Pull request merged**
